Add narrative documentation for NIO's concurrency bridges

FranzBusch opened the pull request:
# Motivation
We started to add bridges from NIO to Swift Concurrency and we should provide some narrative documentation for our users how to adopt them.

# Modification
Adds a new article that explains the NIO concurrency bridges.

FranzBusch commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+to close the channel once both the reading and the writing have finished.
+
+
+This is how you can wrap an existing channel into a `NIOAsyncChannel`, consume the inbound data and
```
We cannot use the symbol link yet because we still use `spi` for the symbols

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+# NIO and Swift Concurrency
+
+This article explains how to interface between NIO and Swift Concurrency 
+
+NIO was created before native Concurrency support in Swift existed, hence, NIO had to solve
+a few problems that have solutions in the language today. Since, the introduction of Swift Concurrency
+NIO has added numerous features to make the inter-op between NIOs channel eventing system and Swift's
+Concurrency primitives as easy as possible.
```
```suggestion
This article explains how to interface between NIO and Swift Concurrency.

NIO was created before native Concurrency support in Swift existed, hence, NIO had to solve
a few problems that have solutions in the language today. Since the introduction of Swift Concurrency,
NIO has added numerous features to make the inter-op between NIO's channel event system and Swift's
Concurrency primitives as easy as possible.
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+
+### EventLoopFuture bridges
+
+The first bridges, that NIO introduced were adding methods on ``EventLoopFuture`` and ``EventLoopPromise``
```
```suggestion
The first bridges that NIO introduced added methods on ``EventLoopFuture`` and ``EventLoopPromise``
```

glbrntt replied:
```suggestion
The first bridges that NIO introduced were methods on ``EventLoopFuture`` and ``EventLoopPromise``
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+The first bridges, that NIO introduced were adding methods on ``EventLoopFuture`` and ``EventLoopPromise``
+to bridge to and from Concurrency. Namely these bridges are ``EventLoopFuture/get()`` and ``EventLoopPromise/completeWithTask(_:)``.
+
+> Warning: The future ``EventLoopFuture/get()`` method does not support task cancellation.
```
```suggestion
> Warning: ``EventLoopFuture/get()`` does not support task cancellation.
```

FranzBusch replied:
I did add the words around because docc renders this as just `get()` which reads very weird.

gjcairo replied:
Ah I see. Okay, then it's fine - with the word `future` though, did you mean `ELF`?

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+### Channel bridges
+
+The ``EventLoopFuture`` and ``EventLoopPromise`` bridges already allow async code to interact with
+some of NIO's types. However, they only work where we need to bridge request-response like interfaces.
```
```suggestion
some of NIO's types. However, they only work where we need to bridge request-response-like interfaces.
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+``NIOAsyncWriter/yield(contentsOf:)`` method.
+
+
+> Important: Everything below this is currently not public API but can be test it by using `@_spi(AsyncChannel) import`.
```
```suggestion
> Important: Everything below this is currently not public API but can be tested by using `@_spi(AsyncChannel) import`.
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+To solve the problem of protocol negotiation we had to introduce a new ``ChannelHandler`` protocol called
+`NIOProtocolNegotiationHandler`. This protocol requires a single future property `NIOProtocolNegotiationHandler/protocolNegotiationResult`
+that is completed once the handler is finished with protocol negotiation. In the successful case,
+the future can either indicate that protocol negotiation is fully done by returning `NIOProtocolNegotiationResult/finished(_:)``or
```
```suggestion
the future can either indicate that protocol negotiation is fully done by returning `NIOProtocolNegotiationResult/finished(_:)` or
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+> Important: Make sure to use discarding task groups which are automatically reaping finished child tasks.
+Normal task groups will result in a memory leak since they do not reap their child tasks automatically.
+
+The above works create for channels that do not have protocol negotiation handlers in their pipeline
```
```suggestion
The above code works for channels that do not have protocol negotiation handlers in their pipeline
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
````diff
@@ -0,0 +1,244 @@
+```
+
+Next, we have to setup our bootstrap. We are adding a `NIOTypedApplicationProtocolNegotiationHandler`
+to each child channel's pipeline This handler is listening to user inbound events of the type `TLSUserEvent`
````
```suggestion
to each child channel's pipeline. This handler is listening to user inbound events of the type `TLSUserEvent`
```

Lukasa replied:
```suggestion
to each child channel's pipeline This handler listens for user inbound events of the type `TLSUserEvent`
```

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+    for try await negotiationResult in serverChannel.inboundStream {
+        group.addTask {
+            do {
+                switch connectionChannel {
```
```suggestion
                switch negotiationResult {
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+
+NIO was created before native Concurrency support in Swift existed, hence, NIO had to solve
+a few problems that have solutions in the language today. Since, the introduction of Swift Concurrency
+NIO has added numerous features to make the inter-op between NIOs channel eventing system and Swift's
```
```suggestion
NIO has added numerous features to make the interop between NIO's `Channel` eventing system and Swift's
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+### EventLoopFuture bridges
+
+The first bridges, that NIO introduced were adding methods on ``EventLoopFuture`` and ``EventLoopPromise``
+to bridge to and from Concurrency. Namely these bridges are ``EventLoopFuture/get()`` and ``EventLoopPromise/completeWithTask(_:)``.
```
```suggestion
to bridge to and from Concurrency. These bridges are ``EventLoopFuture/get()`` and ``EventLoopPromise/completeWithTask(_:)``.
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
````diff
@@ -0,0 +1,244 @@
+let result = try await promise.futureResult.get()
+```
+
+> Note: The `completeWithTask` method is creating an unstructured task under the hood.
````
```suggestion
> Note: The `completeWithTask` method creates an unstructured task under the hood.
```

glbrntt replied:
Is it worth noting that the `Task` is returned to the caller as a `@discardableResult` so they can e.g. cancel if necessary?

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+
+The ``EventLoopFuture`` and ``EventLoopPromise`` bridges already allow async code to interact with
+some of NIO's types. However, they only work where we need to bridge request-response like interfaces.
+On the other hand, NIO's' ``Channel`` type contains a ``ChannelPipeline`` which can be roughly 
```
```suggestion
On the other hand, NIO's ``Channel`` type contains a ``ChannelPipeline`` which can be roughly 
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+The above types are used to bridge both the read and write side of a ``Channel`` into Swift Concurrency.
+This can be done by wrapping a ``Channel`` via the `NIOAsyncChannel/init(synchronouslyWrapping:backpressureStrategy:isOutboundHalfClosureEnabled:inboundType:outboundType:)`
+initializer. Under the hood, this initializer adds two channel handlers to the end of the channel's pipeline.
+These handlers bridge the read and write side of the channel. Additionally, the handlers are working together
```
```suggestion
These handlers bridge the read and write side of the channel. Additionally, the handlers work together
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+
+### Async bootstrap
+
+The `NIOAsyncChannel` alone would already allow you to wrap your channels and use them from Concurrency;
```
```suggestion
The `NIOAsyncChannel` alone allows you to wrap your channels and use them from Concurrency;
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+### Async bootstrap
+
+The `NIOAsyncChannel` alone would already allow you to wrap your channels and use them from Concurrency;
+however, you must be very careful at what point you wrap your channel otherwise you might lose some reads.
```
I think you need to elaborate on this point if you're going to bring it up here. Why might you lose reads? Under what circumstances?

glbrntt replied:
Presumably using an async bootstrap helps you to avoid this so having this at the end of the previous section would sets up the reader nicely for this section nicely.

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+however, you must be very careful at what point you wrap your channel otherwise you might lose some reads.
+Furthermore, with the introduction of `NIOAsyncChannel` we also recommend reevaluating where your
+business logic lives and thinking about moving it from channel handlers into Concurrency tasks. A good split is
+to put network protocol logic into channel handlers and business logic into tasks.
```
This paragraph feels weird to put in the section about bootstraps. Maybe a more general guidance principle?

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+#### Protocol Negotiation
+
+With the introduction of `NIOAsyncChannel` we are introducing concrete types to the channel's pipeline. The underlying
+bridging handlers of the `NIOAsyncChannel` are generic and unwrap the inbound and wrap the outbound data.
```
```suggestion
bridging handlers of the `NIOAsyncChannel` are generic and automatically handle wrapping and unwrapping the outbound and inbound data.
```

Lukasa commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+This means that once you wrap a ``Channel`` into a `NIOAsyncChannel` the inbound and outbound types
+need to be known. Moreover, these types are not allowed to change anymore otherwise it will result in
+runtime crashes.
+To solve the problem of protocol negotiation we had to introduce a new ``ChannelHandler`` protocol called
```
```suggestion
To solve the problem of protocol negotiation we introduced a new ``ChannelHandler`` protocol called
```

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+            group.addTask {
+                do {
+                    for try await inboundData in connectionChannel.inboundStream {
+                        try await  connectionChannel.outboundWriter.write(inboundData)
```
```suggestion
                        try await connectionChannel.outboundWriter.write(inboundData)
```

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+        childChannelOutboundType: ByteBuffer.self
+    )
+
+    try await withThrowingDiscardingTaskGroup { group in
```
This whole block is over indented

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+In the above code, we are bootstrapping a new TCP server which we assign to `serverChannel`. The type of 
+`serverChannel` is a `NIOAsyncChannel` of `NIOAsyncChannel`. This is due to the fact that each
+inbound connection gets its own separate child channel.
+Afterward, we are handling each inbound connection in a separate child task and echo the data back.
```
```suggestion
Afterwards, we handle each inbound connection in separate child tasks and echo the data back.
```

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+to put network protocol logic into channel handlers and business logic into tasks.
+To solve the above timing problems and guide users into the Concurrency based solution the various bootstraps
+gained new methods that directly return `NIOAsyncChannel`s.
+Before diving into how these new bootstrap methods work there is one more problem that needs to be called out -
+ Protocol Negotiation.
```
I think the protocol negotiation should come after the example bootstrap below. The doc is building up from simple to complicated so should introduce new concepts only when they're necessary.

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+with protocol negotiation.
+
+First, we have to define our negotiation result. For this example, we are negotiating between a
+`String` based and `UInt8` based channel. Additionally, we also need an error that we can throw
```
Would be nice if we could use a real example here; e.g. HTTP1 vs HTTP2

FranzBusch replied:
Agreed, but we don't have those yet. Might be worth to swing around later

glbrntt replied:
I think that's probably fine; `AsyncChannel` is still SPI and we're going to do H2 soon.

FranzBusch replied:
Discussed offline: Since we do not know the type that HTTP2 will use for its async abstraction we decided not to showcase it here.

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+                switch connectionChannel {
+                case .string(let channel):
+                    for try await value in channel.inboundStream {
+                        continuation.yield(.string(value))
```
Where did the continuation come from? Is this meant to be echo-ing back the value instead?

glbrntt commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,244 @@
+
+### EventLoopFuture bridges
+
+The first bridges, that NIO introduced were adding methods on ``EventLoopFuture`` and ``EventLoopPromise``
```
The first half of this doc uses the word "bridge" ... a lot.

FranzBusch commented:
@gjcairo @Lukasa @glbrntt Thanks for the great feedback. I fixed all the nits and rearranged some sections to have a better story line. I also added a general guidance section in the end. Right now that only contains guidance around business logic vs protocol logic but we can extend it in the future with guidance around perf as well

gjcairo commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -227,18 +235,33 @@ try await withThrowingDiscardingTaskGroup { group in
+
+#### Where should your code live?
+
+Before the introduction of Swift Concurrency both, implementations of network protocols and business logic,
```
```suggestion
Before the introduction of Swift Concurrency, both implementations of network protocols and business logic
```

gjcairo reviewed (commented):
Looking good - left one more comment.

dnadoba commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,267 @@
+try await withThrowingDiscardingTaskGroup { group in
+    for try await negotiationResult in serverChannel.inboundStream {
+        group.addTask {
+            do {
```
Is there a `catch` missing or why do we use `do` here?

FranzBusch replied:
Good catch! Pun intended

adam-fowler commented on Sources/NIOCore/Docs.docc/swift-concurrency.md:
```diff
@@ -0,0 +1,269 @@
+    protocolNegotiationHandlerType: NIOTypedApplicationProtocolNegotiationHandler<NegotiationResult>.self
+)
+
+try await withThrowingDiscardingTaskGroup { group in
```
Is it worth mentioning `withThrowingDiscardingTaskGroup` is not available in macOS version of Swift 5.8, but is available in Linux?

FranzBusch replied:
I don't think that is the right place to mention this. This should be covered by the documentation of `withThrowingDiscardingTaskGroup`.

FranzBusch commented:
Merging over the nightly failure which is an allocation regression
